type_info.go

```diff
@@ -0,0 +1,414 @@
+		reflect.String:
+		return true
+	case reflect.Struct:
+		for i := 0; i < rt.NumField(); i++ {
+			if rt.Field(i).IsExported() && !canConvertToDuckdb(rt.Field(i).Type) {
+				return false
+			}
+		}
+		return true
+	case reflect.Array, reflect.Slice:
+		elemt := rt.Elem()
+		return canConvertToDuckdb(elemt)
+	case reflect.Map:
+		keyt := rt.Key()
+		elemt := rt.Elem()
+		return canConvertToDuckdb(elemt) && canConvertToDuckdb(keyt)
+	case reflect.Pointer, reflect.Interface:
+		return canConvertToDuckdb(rt.Elem())
+	// This case should never be reached
+	default:
+		return false
+	}
+}
+
+func tryGetDuckdbTypeFromValue(rt reflect.Type) (TypeInfo, error) {
+	switch rt {
+	case reflect.TypeOf(time.Time{}):
+		return NewTypeInfo(Type(TYPE_TIMESTAMP_NS))
+	case reflect.TypeOf(UUID{}):
+		return NewTypeInfo(Type(TYPE_UHUGEINT))
+	case reflect.TypeOf([]byte{}):
+		return NewTypeInfo(Type(TYPE_UUID))
+	}
+	switch rt.Kind() {
+	// Invalid types
+	case reflect.Chan, reflect.Func, reflect.UnsafePointer, reflect.Int, reflect.Uint, reflect.Uintptr, reflect.Complex64, reflect.Complex128:
+		return nil, unsupportedTypeError(rt.String())
+	// Valid types
+	case reflect.Bool:
+		return NewTypeInfo(Type(TYPE_BOOLEAN))
+	case reflect.Int8:
+		return NewTypeInfo(Type(TYPE_TINYINT))
+	case reflect.Int16:
+		return NewTypeInfo(Type(TYPE_SMALLINT))
+	case reflect.Int32:
+		return NewTypeInfo(Type(TYPE_INTEGER))
+	case reflect.Int64:
+		return NewTypeInfo(Type(TYPE_BIGINT))
+	case reflect.Uint8:
+		return NewTypeInfo(Type(TYPE_UTINYINT))
+	case reflect.Uint16:
+		return NewTypeInfo(Type(TYPE_USMALLINT))
+	case reflect.Uint32:
+		return NewTypeInfo(Type(TYPE_UINTEGER))
+	case reflect.Uint64:
+		return NewTypeInfo(Type(TYPE_UBIGINT))
+	case reflect.Float32:
+		return NewTypeInfo(Type(TYPE_FLOAT))
+	case reflect.Float64:
+		return NewTypeInfo(Type(TYPE_DOUBLE))
+	case reflect.String:
+		return NewTypeInfo(Type(TYPE_VARCHAR))
+	case reflect.Struct:
+		fields := make([]StructEntry, 0, rt.NumField())
+		for i := 0; i < rt.NumField(); i++ {
+			if rt.Field(i).IsExported() {
+				field := rt.Field(i)
+				typ, err := tryGetDuckdbTypeFromValue(field.Type)
+				if err != nil {
+					return nil, err
+				}
+
+				entry, err := NewStructEntry(typ, field.Name)
+				fields = append(fields, entry)
+			}
+		}
+		return NewStructInfo(fields[0], fields[1:]...)
+	case reflect.Slice:
+		elemt := rt.Elem()
+		t, err := tryGetDuckdbTypeFromValue(elemt)
+		if err != nil {
+			return nil, err
+		}
+		return NewListInfo(t)
+	case reflect.Map:
+		keyt := rt.Key()
+		kt, err := tryGetDuckdbTypeFromValue(keyt)
+		if err != nil {
+			return nil, err
+		}
+		elemt := rt.Elem()
+		vt, err := tryGetDuckdbTypeFromValue(elemt)
+		if err != nil {
+			return nil, err
+		}
+		return NewMapInfo(kt, vt)
+	case reflect.Pointer, reflect.Interface:
+		return tryGetDuckdbTypeFromValue(rt.Elem())
+	// This case should never be reached
+	default:
+		return nil, unsupportedTypeError(rt.String())
+	}
+}
```

Comment:
I'll leave this as a comment so that we can link to the section in https://github.com/marcboeker/go-duckdb/pull/272.